### PR TITLE
Fix java.lang.StringIndexOutOfBoundsException when unexpected chars in Json string

### LINKED ISSUE #32124: Better error message for worng jsons
Currently it throws `java.lang.StringIndexOutOfBoundsException` for json with trailing commas.
Please give a better error message.

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #32124 

With this fix proper error messages will be given for the above invalid Json string cases.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by warunalakshitha** [comment]
Shall we check why codecov patch coverage is 25%?

**Review by warunalakshitha — COMMENTED** [review]
Shall we check why codecov patch coverage is 25%?

File: langlib/langlib-test/src/test/resources/test-src/valuelib_fromJson_test.bal [hunk 1]
```diff
@@ -828,6 +828,29 @@ function testFromJsonWithTypeToUnionOfTypeReference() {
...
 }
 
+function testFromJsonStringWithUnexpectedChars() {
+    string s1 = "{\"a\":}";
+    string s2 = "[1, 2,]";
+    string s3 = "{\"a\":1,}";
+    string s4 = "{\"a\": [1, 2,]}";
+
+    error err = <error> s1.fromJsonStringWithType(json);
+    assertEquality(<string> checkpanic err.detail()["message"], "expected 'a field value' at line: 1 column: 6");
```
Comment (Nadeeshan96) [inline langlib/langlib-test/src/test/resources/test-src/valuelib_fromJson_test.bal:838]: It is better if we don't have quotes around 'a field value' and 'an array element' (for characters it is okay). i.e. `"expected a field value at line: 1 column: 6"`, `"expected an array element at line: 1 column: 9"`. Shall we change the `StateMachine.throwExpected` method to support that?

Comment (gabilang) [reply langlib/langlib-test/src/test/resources/test-src/valuelib_fromJson_test.bal:838]: +1

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonParser.java [hunk 2]
```diff
@@ -1122,8 +1126,7 @@ public State transition(StateMachine sm, char[] buff, int i, int count) throws J
                         continue;
                     }
                     this.reset(sm);
-                    StateMachine.throwExpected("hexadecimal value of an unicode character");
-                    break;
+                    throw new JsonParserException("expected hexadecimal value of an unicode character");
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonParser.java:1129]:
```suggestion
                    throw new JsonParserException("expected the hexadecimal value of a unicode character");
```
for `unicode` we should use `a` article as per the beginning sound of the word.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonParser.java [hunk 3]
```diff
@@ -1249,7 +1252,7 @@ public State transition(StateMachine sm, char[] buff, int i, int count) throws J
                             }
                             break;
                         default:
-                            StateMachine.throwExpected("escaped characters");
+                            throw new JsonParserException("expected escaped characters");
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonParser.java:1255]: Shall we add a small test to cover this as well?

Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/JsonParser.java:1255]: Added with https://github.com/ballerina-platform/ballerina-lang/pull/40396/commits/909ceb0ff7ff7aca1c4f263117bb358f1d667498

**Review by Nadeeshan96 — APPROVED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**